lib/Sema/CSGen.cpp

```diff
@@ -3078,6 +3078,12 @@ namespace {
         if (auto *elementExpr = getAsExpr<PackElementExpr>(pack)) {
           packType = CS.getType(elementExpr->getPackRefExpr());
         } else if (auto *elementType = getAsTypeRepr<PackElementTypeRepr>(pack)) {
+          // OpenPackElementType sets types for 'each T' type reprs in
+          // expressions. Some invalid code won't make it there, and
+          // the constraint system won't have recorded a type.
+          if (!CS.hasType(elementType->getPackType()))
+            return Type();
```

Comment:
Github doesn’t like when comments appear directly on commit apparently…

Aborting here is not ideal for closure contexts because it fails the closure without diagnostic, we should record invalid reference or similar fix and produce a placeholder type just like we do for ErrorExprs…

Reply:
This happens when type resolution fails earlier than getting to each T type reprs when they're nested, so a diagnostic has already been produced. I've fixed the other cases where pre-check was failing.

Reply:
I can address your suggestion later, but it'll need a little more work than allowing the type variable to bind to a hole, because this type needs to bind to a pack. This means we'll also need to teach `ShapeOf` constraints to gracefully handle hole types without propagating errors, etc. But for right now, this change fixes a crash-on-invalid, and the type resolution error + `unable to infer closure type` is a big improvement already.

Reply:
> a diagnostic has already been produced

This is a big problem in closure contexts currently because the diagnostic is produced mid-solving and unbeknownst to the solver so it’s really hard to reason about, we have this invalid reference fix which effectively lies that it produced a diagnostic (another point is that code completion depends on solver not failing early more and more now). We should try and avoid adding more cases like that if possible, fixing it separately sounds good to me!

Reply:
Filed https://github.com/apple/swift/issues/64109!